File: SwiftAlgorithm/BOJ/1012.swift
//
//  main.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2022/03/08.
//

//1012 유기농 배추
import Foundation

let dx = [-1, 1, 0, 0], dy = [0, 0, -1, 1]

func bfs(_ x: Int, _ y: Int) {
    var queue: [(x: Int, y: Int)] = [(x, y)]
    var index = 0
    
    while index < queue.count {
        let node = queue[index]
        index += 1
        
        for i in 0..<4 {
            let nx = node.x + dx[i]
            let ny = node.y + dy[i]
            
            guard (0..<n) ~= nx, (0..<m) ~= ny, !visited[nx][ny] else {
                continue
            }
            
            visited[nx][ny] = true
            if map[nx][ny] == 1 {
                queue.append((nx, ny))
            }
        }
    }
}

var testcase = Int(readLine()!)!
var m = 0, n = 0, k = 0
var map: [[Int]] = []
var visited: [[Bool]] = []
for _ in 0..<testcase {
    let mnk = readLine()!.split { $0 == " " }.map { Int($0)! }
    m = mnk[0]; n = mnk[1]; k = mnk[2]
    map = Array(repeating: Array(repeating: 0, count: m), count: n)
    for _ in 0..<k {
        let xy = readLine()!.split { $0 == " " }.map { Int($0)! }
        map[xy[1]][xy[0]] = 1
    }
    
    visited = Array(repeating: Array(repeating: false, count: m), count: n)
    var result = 0
    for i in 0..<n {
        for j in 0..<m {
            if !visited[i][j] && map[i][j] == 1 {
                result += 1
                bfs(i, j)
            }
        }
    }
    
    print(result)
}

File: Python/BOJ/BFS/1012.py
import sys
readl = sys.stdin.readline

def bfs(x, y):
  queue = [(x, y)]
  index = 0
  
  maps[x][y] = 0
  
  while index < len(queue):
    node = queue[index]
    index += 1

    for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
      nx = node[0] + dx
      ny = node[1] + dy

      if not (0 <= nx < n and 0 <= ny < m):
        continue
      if not maps[nx][ny] == 1:
        continue

      maps[nx][ny] = 0
      queue.append((nx, ny))

t = int(readl())
for _ in range(t):
  m, n, k = map(int, readl().split())
  maps = [[0 for _ in range(m)] for _ in range(n)]

  for _ in range(k):
    x, y = map(int, readl().split())
    maps[y][x] = 1

  result = 0
  for i in range(n):
    for j in range(m):
      if maps[i][j] == 1:
        result += 1
        bfs(i, j)
  
  print(result)
